moto/s3/models.py

```diff
@@ -1807,16 +1813,21 @@ def list_object_versions(
         )
         # sort by name, revert last-modified-date
         all_versions.sort(key=lambda r: (r.name, -unix_time_millis(r.last_modified)))
-        last_name = None
-        for version in all_versions:
+
+        def key_or_version_match(ver: FakeKey | FakeDeleteMarker) -> bool:
+            if key_marker is None:
+                return True
+
+            if version_id_marker is None:
+                return key_marker == ver.name
+
+            return key_marker == ver.name and version_id_marker == ver.version_id
```

Comment:
Should this be `key_marker <= ver.name and version_id_marker <= ver.version_id`?

As the docs specifically state:
> Specifies the key **to start with** when listing objects in a bucket.

Reply:
That would work if version_id was sequential, but it is basically a random string. I first tried just adding `version_id_marker <= ver.version_id`, but it produced flaky results depending on the random version_id values.

I opted for itertools.dropwhile here, and the helper function simply needs to signal when the first match is found, using exact matching, and dropwhile will then return the remaining items. Quite suitable for this purpose, I think. Was thinking to implement that version_id "search" manually with a version_marker_found bool, but found dropwhile to be a good fit. Let me know if you have thoughts on it.

The main loop still applies the prefix and delimiter logic, and with the max_keys paging added, it looks to me like it behaves like the real S3 server does.

I'd be happy to add more tests if there are suggestions.

Reply:
Fair enough. Would you mind merging in the latest master? That should fix the linter - if the existing tests all pass, then I'd be happy to merge then.